homework_3/calculator/internal/engine/handler/handler.go

```diff
@@ -0,0 +1,107 @@
+package handler
+
+import (
+	"context"
+	"log"
+
+	domainmodels "code-cadets-2021/homework_3/calculator/internal/domain/models"
+	rabbitmqmodels "code-cadets-2021/homework_3/calculator/internal/infrastructure/rabbitmq/models"
+)
+
+// Handler handles bets received and bets calculated.
+type Handler struct {
+	betRepository BetRepository
+}
+
+// New creates and returns a new Handler.
+func New(betRepository BetRepository) *Handler {
+	return &Handler{
+		betRepository: betRepository,
+	}
+}
+
+// HandleBets handles bets.
+func (h *Handler) HandleBets(
+	ctx context.Context,
+	bets <-chan rabbitmqmodels.Bet,
+) <-chan rabbitmqmodels.BetCalculated {
+	resultingBets := make(chan rabbitmqmodels.BetCalculated)
+
+	go func() {
+		defer close(resultingBets)
+
+		for bet := range bets {
+			log.Println("Processing bet, betId:", bet.Id)
+
+			// Calculate the domain bet based on the incoming bet.
+			domainBet := domainmodels.Bet{
+				Id:                   bet.Id,
+				SelectionId:          bet.SelectionId,
+				SelectionCoefficient: bet.SelectionCoefficient,
+				Payment:              bet.Payment,
+			}
+
+			// Insert the domain bet into the repository.
+			err := h.betRepository.InsertBet(ctx, domainBet)
```

Comment:
It may happen that the bet already exists, in which case it should be updated instead of inserted. Imagine if there was another microservice in this system, e.g. that service is updating `Payment` of the bet. Controller would pick up the change and produce it to `bets` queue, while Calculator would need to update that `Payment` in its local database.
You may need to update `betRepository` interface in order to achieve the desired behaviour (check if the bet already exists in the repository and then act accordingly).

Reply:
What would be the best way to do that? My idea is to add 2 methods to `betRepository`, one that will check if bet already exists in the database and the other that will update the existing bet. Implemented as such, `handler` will check if the bet already exists in the database, if it does it will be updated and otherwise inserted.

Reply:
Yep, that is the best way I can think of.

Reply:
Changes are pushed to the PR.